src/Workspaces/Core/Portable/Storage/CloudCache/ProjectCacheContainerKey.cs

```diff
@@ -0,0 +1,99 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Runtime.CompilerServices;
+using Microsoft.CodeAnalysis.PersistentStorage;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Storage
+{
+    /// <summary>
+    /// Cache of our own internal roslyn storage keys to the equivalent platform cloud cache keys.  Cloud cache keys can
+    /// store a lot of date in them (like their 'dimensions' dictionary.  We don't want to continually recreate these as
+    /// we read/write date to the db.
+    /// </summary>
+    internal class ProjectCacheContainerKey
+    {
+        private static readonly ImmutableSortedDictionary<string, string?> EmptyDimensions = ImmutableSortedDictionary.Create<string, string?>(StringComparer.Ordinal);
+
+        /// <summary>
+        /// Container key explicitly for the project itself.
+        /// </summary>
+        public readonly CloudCacheContainerKey? ContainerKey;
+
+        /// <summary>
+        /// Cache from document green nodes to the container keys we've computed for it. We can avoid computing these
+        /// container keys when called repeatedly for the same documents.
+        /// </summary>
+        private readonly ConditionalWeakTable<TextDocumentState, StrongBox<CloudCacheContainerKey?>> _documentToContainerKey = new();
```

Comment:
Why the conditional weak table here? If the outer thing holding onto this is already doing a CWT on ProjectState, I can't think of a way where keys would ever fall out this table since the ProjectState will already root all of the DocumentStates inside of it.

Reply:
Oh, that's a cool point.  I hadn't considered that.  Will def simplify this.  Thanks!